File: LegacySource/settlement/room/food/orchard/Gui.cs
using System;
using System.Collections.Generic;
using System.Linq;
using settlement.room.food.orchard;
using game;
using init.sprite;
using settlement.room.industry.module;
using settlement.room.main;
using snake2d.util.datatypes;
using snake2d.util.gui;
using snake2d.util.gui.clickable;
using snake2d.util.gui.renderable;
using snake2d.util.sets;
using util.data;
using util.gui.misc;
using util.gui.table;
using util.info;
using util.text;
using view.sett.ui.room;

class Gui : UIRoomModuleImp<Instance, ROOM_ORCHARD>
{
    static CharSequence ¤¤Trees = "¤Trees";
    static CharSequence ¤¤TreesD = "¤Amount of fully grown trees. Only when trees are fully grown will they start producing. Neglected trees will die.";
    private static CharSequence ¤¤TreeNext = "¤Next tree will be grown in:";

    private static CharSequence ¤¤estimated = "¤Estimated Harvest (year)";
    private static CharSequence ¤¤daysToHarvest = "¤Days until harvest";
    private static CharSequence ¤¤baseValue = "¤Capacity";

    private static CharSequence ¤¤HarvestYear = "¤This Year";
    private static CharSequence ¤¤HarvestPrev = "¤Last Year";

    private static CharSequence ¤¤skill = "¤Worker Skill";
    private static CharSequence ¤¤skillD = "¤The average bonus accumulated during the year. This determines the output of the harvest and the growth rate of the trees.";
    private static CharSequence ¤¤skillCurrent = "¤Bonus (current)";
    private static CharSequence ¤¤skillCurrentD = "¤The bonus that is currently being added to the Orchard.";

    private static CharSequence ¤¤chop = "¤Chop";
    private static CharSequence ¤¤chopD = "¤Reset all progress by chopping down the trees and instantly get {0} {1}.";

    Gui(ROOM_ORCHARD s) : base(s)
    {
        D.t(this);
    }

    final Cache cache = new Cache();

    public override void hover(GBox box, Instance i)
    {
        base.hover(box, i);
        box.NL();
        if (!i.blueprintI().constructor.isIndoors)
        {
            box.text(blueprint.constructor.fertility.name());
            box.add(GFORMAT.perc(box.text(), blueprint.constructor.fertility.get(i)));

            box.NL();
        }
        box.text(¤¤estimated);
        box.add(GFORMAT.i(box.text(), (int)cache.output(i)));

        box.space();
    }

    protected override void appendMain(GGrid icons, GGrid text, GuiSection sExtra)
    {
        IndustryResource res = blueprint.industries().get(0).outs().get(0);

        text.add(new GHeader(Dic.¤¤Production));

        GuiSection s = new GuiSection()
        {
            public override void hoverInfoGet(GUI_BOX text)
            {
                text.title(¤¤estimated);
            }
        };

        s.add(res.resource.icon(), 0, 0);

        s.addRightC(4, new GStat()
        {
            public override void update(GText text)
            {
                int am = 0;
                for (int i = 0; i < blueprint.instancesSize(); i++)
                {
                    Instance ins = blueprint.getInstance(i);
                    am += cache.output(ins);
                }
                GFORMAT.i(text, am);
            }
        });

        text.add(s);

        GStaples st = new GStaples(res.history().historyRecords())
        {
            protected override void hover(GBox box, int stapleI)
            {
                int i = res.history().historyRecords() - 1 - stapleI;
                int am = res.history().get(i);
                GText t = box.text();
                DicTime.setDaysAgo(t, i);
                box.add(t);
                box.NL(2);
                box.add(GFORMAT.i(box.text(), am));
            }

            protected override double getValue(int stapleI)
            {
                return res.history().get(res.history().historyRecords() - 1 - stapleI);
            }
        };

        st.body().setWidth(180).setHeight(64);
        text.add(st);
    }

    protected override void appendPanel(GuiSection section, GGrid grid, GuiSection sExtra)
    {
        IndustryResource res = blueprint.industries().get(0).outs().get(0);

        GuiSection s = new GuiSection()
        {
            public override void hoverInfoGet(GUI_BOX text)
            {
                GBox b = (GBox)text;
                Instance ins = getter.get();

                b.textL(¤¤baseValue);
                b.tab(6);
                b.add(GFORMAT.f(b.text(), ins.baseValue));
                b.NL();

                b.textL(Dic.¤¤ProductionRate);
                b.tab(6);
                b.add(GFORMAT.f(b.text(), blueprint.productionData.outs().get(0).rate));
                b.NL();

                b.textL(¤¤skill);
                b.tab(6);
                b.add(GFORMAT.f(b.text(), ins.skill()));
                b.NL();

                b.textLL(Dic.¤¤Total);
                b.tab(6);
                b.add(GFORMAT.f1(b.text(), ins.baseValue * blueprint.productionData.outs().get(0).rate * ins.skill()));
                b.NL();

                b.textL(¤¤Trees);
                b.tab(6);
                b.add(GFORMAT.f(b.text(), (double)ins.trees / ins.treesTotal));
                b.NL();

                b.textLL(¤¤estimated);
                b.tab(6);
                b.add(GFORMAT.f1(b.text(), cache.output(ins)));
                b.NL();

                b.NL(4);

                b.NL(16);
                b.textL(¤¤HarvestYear);
                b.tab(6);
                b.add(GFORMAT.i(b.text(), (int)ins.blueprintI().indus.get(0).outs().get(0).year.get(ins)));
                b.NL();

                b.NL(2);
                b.textL(¤¤HarvestPrev);
                b.tab(6);
                b.add(GFORMAT.i(b.text(), (int)ins.blueprintI().indus.get(0).outs().get(0).yearPrev.get(ins)));
                b.NL();
            }
        };

        s.add(blueprint.productionData.outs().get(0).resource.icon(), 0, 0);
        GStat stat = new GStat()
        {
            public override void update(GText text)
            {
                double am = cache.output(getter.get()) / TIME.years().bitConversion(TIME.days());
                GFORMAT.f0(text, am);
            }
        };

        s.addRightC(6, stat);

        s.body().incrW(64);
        return s;
    }

    private class Cache
    {
        private int upI = -1;
        private int treesTotal;
        private int trees;
        private int daysTillNextTree;
        private int wood;
        private double output;

        private Instance ins;

        public int treesTotal(Instance ins)
        {
            up(ins);
            return treesTotal;
        }

        public int trees(Instance ins)
        {
            up(ins);
            return trees;
        }

        public int daysTillNextTree(Instance ins)
        {
            up(ins);
            return daysTillNextTree;
        }

        public double output(Instance ins)
        {
            up(ins);
            return output;
        }

        private void up(Instance ins)
        {
            if (upI == GAME.updateI() && this.ins == ins)
                return;
            this.ins = ins;
            upI = GAME.updateI();

            wood = 0;
            treesTotal = 0;
            trees = 0;
            daysTillNextTree = int.MaxValue;

            foreach (Coordinate c in ins.body())
            {
                if (!ins.is(c))
                    continue;

                OTile t = blueprint.tile.getM(c.x(), c.y());
                if (t != null)
                {
                    treesTotal++;
                    if (t.state() == t.IBIG)
                        trees++;
                    else
                    {
                        int d = t.state().daysTillGrown();
                        if (d < daysTillNextTree)
                            daysTillNextTree = d;
                    }
                    if (t.state() == t.ISMALL)
                        wood += blueprint.auxRes.amount() / 2;
                    else if (t.state() == t.IBIG || t.state() == t.IDEAD)
                        wood += blueprint.auxRes.amount();
                }
            }

            output = blueprint.time.days * ins.skill() * ins.baseValue * ins.industry().outs().get(0).rate * ins.trees / ins.treesTotal;
        }
    }
}

File: settlement/room/food/farm/Gui.java
package settlement.room.food.farm;

import game.faction.FACTIONS;
import game.time.TIME;
import init.sprite.SPRITES;
import settlement.main.SETT;
import settlement.room.industry.module.IndustryResource;
import settlement.room.industry.module.IndustryUtil;
import settlement.room.main.RoomInstance;
import snake2d.util.datatypes.DIR;
import snake2d.util.gui.GUI_BOX;
import snake2d.util.gui.GuiSection;
import snake2d.util.gui.clickable.CLICKABLE;
import snake2d.util.gui.renderable.RENDEROBJ;
import snake2d.util.misc.ACTION;
import snake2d.util.misc.CLAMP;
import snake2d.util.sets.LISTE;
import snake2d.util.sets.Stack;
import snake2d.util.sprite.text.Str;
import util.data.GETTER;
import util.gui.misc.GBox;
import util.gui.misc.GButt;
import util.gui.misc.GGrid;
import util.gui.misc.GHeader;
import util.gui.misc.GStat;
import util.gui.misc.GText;
import util.gui.table.GStaples;
import util.gui.table.GTableSorter.GTFilter;
import util.gui.table.GTableSorter.GTSort;
import util.info.GFORMAT;
import util.text.D;
import util.text.Dic;
import util.text.DicTime;
import view.main.VIEW;
import view.sett.ui.room.UIRoomBulkApplier;
import view.sett.ui.room.UIRoomModule.UIRoomModuleImp;

class Gui extends UIRoomModuleImp<FarmInstance, ROOM_FARM> {

	private static CharSequence ¤¤estimated = "¤Estimated Harvest (year)";
	private static CharSequence ¤¤daysToHarvest = "¤Days until harvest";
	private static CharSequence ¤¤baseValue = "¤Base Value";
	private static CharSequence ¤¤workValue = "¤Work Value";
	private static CharSequence ¤¤workValueD = "¤Each day of the year a farm needs tending to. Workers failing to tend to all the tiles will lead to low yields.";
	
	private static CharSequence ¤¤HarvestYear = "¤This Year";
	private static CharSequence ¤¤HarvestPrev = "¤Last Year";
	
	private static CharSequence ¤¤skill = "¤Bonus";
	private static CharSequence ¤¤skillD = "¤The average bonus accumulated during the year. This determines the output of the harvest.";
	
	private static CharSequence ¤¤skillCurrent = "¤Bonus (current)";
	private static CharSequence ¤¤skillCurrentD = "¤The bonus that is currently being added to the Farm.";
	
	private static CharSequence ¤¤cycle = "¤Cycle";
	private static CharSequence ¤¤reseed = "¤reseed";
	private static CharSequence ¤¤reseedD = "¤Reseed the farm with another crop. New farm must still be constructed.";
	
	private static CharSequence ¤¤Farmer = "¤Farmers have no storage nearby to store their harvest.";
	
	static {
		D.ts(Gui.class);
	}
	
	Gui(ROOM_FARM s) {
		super(s);
	}
	
	private GuiSection rebuilds = new GuiSection();
	
	@Override
	public void hover(GBox box, FarmInstance i) {
		super.hover(box, i);
		box.NL();
		box.text(blueprint.constructor.fertility.name());
		box.add(GFORMAT.perc(box.text(), blueprint.constructor.fertility.get(i)));
		
		box.NL();
		box.text(¤¤estimated);
		box.add(GFORMAT.i(box.text(), (int)Math.ceil(Util.prospect(i))));
		
		box.space();
	}
	
	@Override
	protected void problem(FarmInstance i, Stack<Str> free, LISTE<CharSequence> errors, LISTE<CharSequence> warnings) {
		if (i.tData.shouldStore() && i.storeTimeout) {
			warnings.add(¤¤Farmer);
		}
		super.problem(i, free, errors, warnings);
	}
	
	@Override
	protected void appendMain(GGrid icons, GGrid text, GuiSection sExtra) {
		
		IndustryResource res = blueprint.industries().get(0).outs().get(0);
		
		text.add(new GHeader(Dic.¤¤Production));
		
		GuiSection s = new GuiSection() {
			
			@Override
			public void hoverInfoGet(GUI_BOX text) {
				text.title(¤¤estimated);
				
			};
			
		};
		
		s.add(res.resource.icon(), 0, 0);
		
		s.addRightC(4, new GStat() {
			
			@Override
			public void update(GText text) {
				int am = 0;
				for (int i = 0; i < blueprint.instancesSize(); i++) {
					FarmInstance ins = blueprint.getInstance(i);
					am += Util.prospect(ins);
				}
				GFORMAT.i(text, am);
			}
		});
		
		s.addRightC(64, new GStat() {
			
			@Override
			public void update(GText text) {
				int prev = 0;
				int am = 0;
				for (int i = 0; i < blueprint.instancesSize(); i++) {
					FarmInstance ins = blueprint.getInstance(i);
					prev += Util.prevHarvest(ins);
					am += Util.prospect(ins);
				}
				GFORMAT.iIncr(text, am-prev);
			}
		});
		
		text.add(s);
		
		GStaples st = new GStaples(res.history().historyRecords()) {
			
			@Override
			protected void hover(GBox box, int stapleI) {
				int i = res.history().historyRecords()-1-stapleI;
				int am = res.history().get(i);
				GText t = box.text();
				DicTime.setDaysAgo(t, i);
				box.add(t);
				box.NL(2);
				box.add(GFORMAT.i(box.text(), am));
			}
			
			@Override
			protected double getValue(int stapleI) {
				return res.history().get(res.history().historyRecords()-1-stapleI);
			}
		};
		

		st.body().setWidth(180).setHeight(64);
		text.add(st);
		
	}

	@Override
	protected void appendPanel(GuiSection section, GGrid grid, GETTER<FarmInstance> getter, int x1, int y1) {
		
		GuiSection s = new GuiSection();
		
		s.add(prod(getter));
		
		RENDEROBJ o = new GStat() {
			@Override
			public void update(GText text) {
				DicTime.setDays(text, blueprint.time.daysToHarvest());
			}
		}.hh(SPRITES.icons().s.clock).hoverInfoSet(¤¤daysToHarvest);
		s.addRightC(32, o);
		s.body().incrW(64);
		
		s.addRelBody(4, DIR.N, new GHeader(Dic.¤¤Production));
		
		section.addRelBody(8, DIR.S, s);
		
		s = new GuiSection();
		
//		s.add(new GStat() {
//			
//			@Override
//			public void update(GText text) {
//				GFORMAT.perc(text, getter.get().tData.fertility());
//			}
//		}.hv(Fertility.¤¤name, blueprint.constructor.fertility.desc()));
			
		
		s.addRightC(32, new GStat() {
			
			@Override
			public void update(GText text) {
				GFORMAT.perc(text, getter.get().tData.work());
			}
			
			@Override
			public void hoverInfoGet(GBox b) {
				b.title(¤¤workValue);
				b.text( ¤¤workValueD);
				b.NL();
				b.textLL(DicTime.¤¤Today);
				b.tab(4);
				b.add(GFORMAT.perc(b.text(), CLAMP.d(getter.get().tData.workday(), 0, 1)));
			};
			
		}.hv(¤¤workValue));
		
		s.addRightC(32, new GStat() {
			
			@Override
			public void update(GText text) {
				GFORMAT.f1(text, getter.get().tData.skill());
			}
			
			@Override
			public void hoverInfoGet(GBox b) {
				b.title(¤¤skill);
				b.text(¤¤skillD);
				b.NL(8);
				b.textLL(Dic.¤¤Value);
				b.tab(6);
				b.add(GFORMAT.f1(b.text(), getter.get().tData.skill()));
				
				b.sep();
				b.textLL(¤¤skillCurrent);
				b.NL();
				b.text(¤¤skillCurrentD);
				b.NL();
				IndustryUtil.hoverBoosts(b, 1.0, getter.get().industry(), getter.get().industry().bonus(), getter.get(), 1);

			};
			
		}.hv(¤¤skill));
		
		section.addRelBody(16, DIR.S, s);
		s = new GuiSection();
		
		
		s.addRightC(32, new GStat() {
			
			@Override
			public void update(GText text) {
				text.add(getter.get().tData.cName());
			}
		}.hv(¤¤cycle));
		section.addRelBody(4, DIR.S, s);
		
		ACTION rebuild = new ACTION() {
			@Override
			public void exe() {
				rebuilds = new GuiSection();
				int i = 0;
				for (ROOM_FARM f : SETT.ROOMS().FARMS) {
					if (f.isAvailable(SETT.ENV().climate())) {
						CLICKABLE ss = new GButt.ButtPanel(f.iconBig()) {
							
							@Override
							protected void clickA() {
								if (f.reqs.passes(FACTIONS.player()) && f != getter.get().blueprintI()) {
									VIEW.inters().popup.close();
									getter.get().changeTo(f);
									
								}
								
							};
							
							@Override
							protected void renAction() {
								activeSet(f.reqs.passes(FACTIONS.player()) && f != getter.get().blueprintI() && f.constructor.isIndoors == getter.get().blueprintI().constructor.isIndoors);
							};
							
						}.hoverSet(f.info);
						rebuilds.add(ss, (i%6)*ss.body().width(), (i/6)*ss.body().height());
						i++;
					}
				}
				
			}
		};
		{
			rebuild.exe();
			SETT.addGeneratorHook(rebuild);
		}
		
		CLICKABLE c = new GButt.ButtPanel(¤¤reseed) {
			@Override
			protected void clickA() {
				VIEW.inters().popup.show(rebuilds, this);
				super.clickA();
			}
		}.pad(8, 4).hoverInfoSet(¤¤reseedD);
		
		section.addRelBody(16, DIR.S, c);
	
		
		
	}
	
	private RENDEROBJ prod(GETTER<FarmInstance> getter) {
		GuiSection s = new GuiSection() {
			
			@Override
			public void hoverInfoGet(GUI_BOX text) {
				GBox b = (GBox) text;
				FarmInstance ins = getter.get();

				b.textL(¤¤baseValue);
				b.tab(6);
				b.add(GFORMAT.f(b.text(), Util.base(ins)));
				b.NL();
				
				b.textL(Dic.¤¤ProductionRate);
				b.tab(6);
				b.add(GFORMAT.f(b.text(), blueprint.productionData.outs().get(0).rate));
				b.NL();

				b.textL(¤¤workValue);
				b.tab(6);
				b.add(GFORMAT.f1(b.text(), ins.tData.work()));
				b.NL();
				
				b.textL(¤¤skill);
				b.tab(6);
				b.add(GFORMAT.f1(b.text(), ins.tData.skill()));
				b.NL();
				
				
				b.NL(4);
				
				b.textLL(¤¤estimated);
				b.tab(6);
				b.add(GFORMAT.f1(b.text(), Util.prospect(ins)));
				b.NL();
				
				b.NL(16);
				b.textL(¤¤HarvestYear);
				b.tab(6);
				b.add(GFORMAT.i(b.text(), (int)ins.blueprintI().indus.get(0).outs().get(0).year.get(ins)));
				b.NL();
				
				b.NL(2);
				b.textL(¤¤HarvestPrev);
				b.tab(6);
				b.add(GFORMAT.i(b.text(), (int)ins.blueprintI().indus.get(0).outs().get(0).yearPrev.get(ins)));
				b.NL();
				
			}
			
		};
		
		s.add(blueprint.crop.resource.icon(), 0, 0);
		GStat stat = new GStat() {
			
			@Override
			public void update(GText text) {
				double am = Util.prospect(getter.get())/TIME.years().bitConversion(TIME.days());
				GFORMAT.f0(text, am);
			}
		};
		
		s.addRightC(6, stat);
		
		s.body().incrW(64);
		return s;
	}
	
	@Override
	protected void appendTableFilters(LISTE<GTFilter<RoomInstance>> filters, LISTE<GTSort<RoomInstance>> sorts,
			LISTE<UIRoomBulkApplier> appliers) {
	}


}
